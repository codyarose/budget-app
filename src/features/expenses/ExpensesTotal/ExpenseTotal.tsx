import React, { FC } from 'react'
import numeral from 'numeral'

interface Props {
	count: number
	total: number
}

const ExpenseTotal: FC<Props> = ({ count, total }) => {
	const expensePlural = count === 1 ? 'expense' : 'expenses'
	const formattedTotal = numeral(total / 100).format('$0,0.00')
	return (
		<>
			{!!count && (
				<div>
					Showing {count} {expensePlural} totalling {formattedTotal}
				</div>
			)}
		</>
	)
}

export default ExpenseTotal
