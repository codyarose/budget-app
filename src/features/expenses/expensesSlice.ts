import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { v4 as uuid } from 'uuid'
import { Moment } from 'moment'

import { RootState } from '../../app/store'
import moment from 'moment'

export interface Expense {
	id: string
	description: string
	note: string
	amount: number
	createdAt: Moment | number
}

export type ExpenseData = Omit<Expense, 'id'>

export const initialState: Expense[] = []

export const expensesSlice = createSlice({
	name: 'expenses',
	initialState,
	reducers: {
		addExpense: (state, action: PayloadAction<Partial<ExpenseData>>) => {
			const {
				description = '',
				note = '',
				amount = 0,
				createdAt = 0,
			} = action.payload
			state.push({
				id: uuid(),
				description,
				note,
				amount,
				createdAt,
			})
		},
		removeExpense: (state, action: PayloadAction<string>) => {
			const id = action.payload
			const index = state.findIndex((expense) => expense.id === id)
			if (index === -1) {
				return state
			} else {
				state.splice(index, 1)
			}
		},
		editExpense: (
			state,
			action: PayloadAction<{
				editedExpense: Partial<ExpenseData>
				id: string
			}>,
		) => {
			const { editedExpense, id } = action.payload
			return state.map((expense) => {
				if (expense.id === id) {
					return {
						...expense,
						...editedExpense,
					}
				} else {
					return expense
				}
			})
		},
	},
})

export const { addExpense, removeExpense, editExpense } = expensesSlice.actions

export const selectExpenses = (state: RootState): Expense[] => state.expenses

export const selectVisibleExpenses = (state: RootState): Expense[] => {
	const expenses = state.expenses
	const { startDate, endDate, text, sortBy } = state.filter

	return expenses
		.filter((expense) => {
			const createdAtMoment = moment(expense.createdAt)
			const momentStartDate = moment(startDate)
			const momentEndDate = moment(endDate)

			const startDateMatch = startDate
				? momentStartDate.isSameOrBefore(createdAtMoment, 'day')
				: true
			const endDateMatch = endDate
				? momentEndDate.isSameOrAfter(createdAtMoment, 'day')
				: true
			const textMatch = expense.description
				.toLowerCase()
				.includes(text.toLowerCase())

			return startDateMatch && endDateMatch && textMatch
		})
		.sort((a, b) => {
			if (sortBy === 'date') {
				return a.createdAt < b.createdAt ? 1 : -1
			} else if (sortBy === 'amount') {
				return a.amount < b.amount ? 1 : -1
			} else {
				return 1
			}
		})
}

export const selectTotal = (state: RootState): number => {
	const expenses = selectVisibleExpenses(state)

	if (!expenses.length) return 0

	return expenses.reduce((acc, expense) => {
		return acc + expense.amount
	}, 0)
}

export default expensesSlice.reducer
