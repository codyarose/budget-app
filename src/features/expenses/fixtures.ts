export default [
	{
		id: '123abc',
		description: 'Title',
		note: '',
		amount: 140500,
		createdAt: 1598763926001,
	},
	{
		id: '456def',
		description: 'Blah',
		note: 'This is a note',
		amount: 2650,
		createdAt: 1598763926009,
	},
	{
		id: '789ghi',
		description: 'Another one',
		note: 'Text goes here',
		amount: 340098,
		createdAt: 1598763926345,
	},
]
