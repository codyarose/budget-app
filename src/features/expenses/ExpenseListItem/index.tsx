import React, { FC } from 'react'
import { Link } from 'react-router-dom'

import { Expense } from '../expensesSlice'

const ExpenseListItem: FC<Expense> = ({
	description,
	amount,
	createdAt,
	note,
	id,
}) => {
	return (
		<div>
			<Link to={`/edit/${id}`}>
				<h4>{description}</h4>
			</Link>
			<p>
				{amount} - {createdAt}
			</p>
			{note && <p>{note}</p>}
		</div>
	)
}

export default ExpenseListItem
