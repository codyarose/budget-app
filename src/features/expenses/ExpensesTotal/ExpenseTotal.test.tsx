import React from 'react'
import { shallow, ShallowWrapper } from 'enzyme'

import ExpenseTotal from '.'
import expenses from '../fixtures'

let props: { count: number; total: number }, wrapper: ShallowWrapper
beforeEach(() => {
	props = {
		count: 0,
		total: 0,
	}
	wrapper = shallow(<ExpenseTotal {...props} />)
})

describe('ExpenseTotal', () => {
	it('should not render if there are no expenses', () => {
		expect(wrapper).toMatchSnapshot()
	})

	it('should render correct cound and total with one expense', () => {
		const expense = [expenses[0]]

		wrapper.setProps({
			count: expense.length,
			total: expense[0].amount,
		})
		expect(wrapper).toMatchSnapshot()
	})

	it('should render correct cound and total with multiple expenses', () => {
		const total = expenses.reduce((acc, expense) => {
			return acc + expense.amount
		}, 0)

		wrapper.setProps({
			count: expenses.length,
			total,
		})
		expect(wrapper).toMatchSnapshot()
	})
})
