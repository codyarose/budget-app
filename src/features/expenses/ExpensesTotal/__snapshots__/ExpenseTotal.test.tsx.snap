// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseTotal should not render if there are no expenses 1`] = `<Fragment />`;

exports[`ExpenseTotal should render correct cound and total with multiple expenses 1`] = `
<Fragment>
  <div>
    Showing 
    3
     
    expenses
     totalling 
    $4,832.48
  </div>
</Fragment>
`;

exports[`ExpenseTotal should render correct cound and total with one expense 1`] = `
<Fragment>
  <div>
    Showing 
    1
     
    expense
     totalling 
    $1,405.00
  </div>
</Fragment>
`;
